README.md

```diff
@@ -1,2 +1,202 @@
+}
+
+message StopRequest {
+  int32 job_id = 1;
+}
+
+message StopResponse {
+  string message = 1;
+}
+
+message StatusRequest {
+  int32 job_id = 1;
+}
+
+message StatusResponse {
+  int32 job_id = 1;
+  string status = 2;
+  bool exited = 3;
+  int32 exit_code = 4;
+}
+
+message OutputRequest {
+  int32 job_id = 1;
+}
+
+message OutputResponse {
+  repeated string output_line = 1;
+}
+```
+1. **`Start`** will send various commands batched into a single RPC to the server and server will create `Job` for each command and send back a batch of status response.
+1. **`Stop`** will and RPC with a single job id to a server and server will stop the specified job if authorization is successful. A response will be sent back with an appropriate message.
+`- successfully stopped job with id <id>`
+`- error: job id not found`
+If authorization fails for a running job, then instead of admitting that "such job exists but failed to authorize for stopping it", it is better to simply to say `job id not found`. Similar to how entering wrong password for a valid email tells that `username or password is incorrect` rather than saying that `password is incorrect`.
+
+1. **`Status`**, will request for currrert status/state of a job and server will send back a response containing its state, exit code and some human readable message (`scheduled`, `running`, `completed`, etc.)
+1. **`Output`** will request for output of a single job and server will stream its output back line by line.
+Note: It will always stream from the start of the job; no matter how many times requested. If job is complete, it will stream all the way to the end. If job is still running, it will wait for new data to arrive until server completes the stream. Behaviour will be similar to of `tail -f`.
+To optimize the performance a little, `OutputResponse` will consist of a string slice `[]string`. Server will try to batch up to `N` lines of output (`N` can be any reasonable number: 10, 20, 30, etc.) and send them with a single RPC response to reduce the number of calls needed over the network to complete the task. If the channel feeding the information does not have `N` lines of output, server will not wait and instead send whatever it has batched into the response.
+
+
+### Authentication
+Communication channels between clients and server will be protected via mutual TLS (mTLS). For simplicity, I will create a root CA and sign certificate for both server and its clients. Clients and the server will use these generated static certificates. Server and clients will authenticate each other by requesting other party's certificate and checking it with CA.
+
+### Authorization Scheme
+For simplicity, each client certificate will embed client's username in `Common Name` field. Server. will extract this information and authorize RPCs on existing jobs if and only if it belongs to that user.
+Server will record each created Job in a map data structure. It will look like this:
+`map[string]map[int]*Job`
```

Comment:
Will the most common lookup for a job be via job ID or via username? I'd expect it would be by ID and the data structure should be optimized for that. My intuition about the most common lookup may be invalid for your implementation, though.

Reply:
Will change the code accordingly.
Implementation changes are not a problem. As of now, Job structure only has id, not username of the client who created it. I will change the storage type to `map[UUID]*Job` and username will be part of the Job. During authorization, Job will be looked up by ID first, then username will be compared to either allow or disallow the operation.